File: platforms/android/states/src/main/java/com/facebook/flipper/nativeplugins/table/TableRow.java
package com.facebook.states.nativeplugins.table;

import com.facebook.states.core.StatesObject;
import com.facebook.states.nativeplugins.components.Sidebar;
import java.util.Map;

public abstract class TableRow {
  public interface Value {
    StatesObject serialize();
  }

  public static class StringValue implements Value {
    private String val;

    public StringValue(String s) {
      this.val = s;
    }

    @Override
    public StatesObject serialize() {
      return new StatesObject.Builder().put("type", "string").put("value", val).build();
    }
  }

  public static class IntValue implements Value {
    private int val;

    public IntValue(int i) {
      this.val = i;
    }

    @Override
    public StatesObject serialize() {
      return new StatesObject.Builder().put("type", "int").put("value", val).build();
    }
  }

  public static class BooleanValue implements Value {
    private boolean val;

    public BooleanValue(boolean i) {
      this.val = i;
    }

    @Override
    public StatesObject serialize() {
      return new StatesObject.Builder().put("type", "boolean").put("value", val).build();
    }
  }

  public static class TimeValue implements Value {
    private long millis;

    public TimeValue(long millis) {
      this.millis = millis;
    }

    @Override
    public StatesObject serialize() {
      return new StatesObject.Builder().put("type", "time").put("value", millis).build();
    }
  }

  public static class DurationValue implements Value {
    private long millis;

    public DurationValue(long millis) {
      this.millis = millis;
    }

    @Override
    public StatesObject serialize() {
      return new StatesObject.Builder().put("type", "duration").put("value", millis).build();
    }
  }

  final String id;
  final Map<Column, ? extends Value> values;
  final Sidebar sidebar;

  public TableRow(String id, Map<Column, ? extends Value> values, Sidebar sidebar) {
    this.id = id;
    this.values = values;
    this.sidebar = sidebar;
  }

  final StatesObject serialize() {
    StatesObject.Builder columnsObject = new StatesObject.Builder();
    for (Map.Entry<Column, ? extends Value> e : values.entrySet()) {
      columnsObject.put(e.getKey().id, e.getValue().serialize());
    }
    columnsObject.put("id", id);
    return new StatesObject.Builder()
        .put("columns", columnsObject.build())
        .put("sidebar", sidebar.serialize())
        .put("id", id)
        .build();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null) {
      return false;
    }
    if (getClass() != o.getClass()) {
      return false;
    }
    return serialize().equals(((TableRow) o).serialize());
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/nativeplugins/table/TableRow.kt
package org.stato.nativeplugins.table

import org.stato.core.StatoObject
import org.stato.nativeplugins.components.Sidebar

abstract class TableRow(val id: String?, val values: Map<Column, Value>?, val sidebar: Sidebar?) {
  interface Value {
    fun serialize(): StatoObject
  }

  class StringValue(private val `val`: String) : Value {

        override fun serialize(): StatoObject {
      return StatoObject.Builder().put("type", "string").put("value", `val`).build()
    }
  }

  class IntValue(private val `val`: Int) : Value {

        override fun serialize(): StatoObject {
      return StatoObject.Builder().put("type", "int").put("value", `val`).build()
    }
  }

  class BooleanValue(private val `val`: Boolean) : Value {

        override fun serialize(): StatoObject {
      return StatoObject.Builder().put("type", "boolean").put("value", `val`).build()
    }
  }

  class TimeValue(private val millis: Long) : Value {

        override fun serialize(): StatoObject {
      return StatoObject.Builder().put("type", "time").put("value", millis).build()
    }
  }

  class DurationValue(private val millis: Long) : Value {

        override fun serialize(): StatoObject {
      return StatoObject.Builder().put("type", "duration").put("value", millis).build()
    }
  }

  internal fun serialize(): StatoObject {
    val columnsObject = (values?: emptyMap()).entries.fold(StatoObject.Builder()) { builder, (key, value) ->
      builder.put(key.id, value.serialize())
    }.put("id", id)


    return StatoObject.Builder()
      .put("columns", columnsObject.build())
      .put("sidebar", sidebar!!.serialize())
      .put("id", id)
      .build()
  }

    override fun equals(other: Any?): Boolean {
      return serialize() == (other as? TableRow)?.serialize()

  }
}
